### ISSUE TITLE
关于ENet-Label-Torch和ERFNet-CULane-PyTorch的在CULane testing set 中的疑问

### ISSUE THREAD

**Issue by Baiwenjing** [cardwing/Codes-for-Lane-Detection#209]
您好，我对2. CULane testing set（F1-measure）的对比结果有一些疑问。列表中，ERFNet-CULane-PyTorch的结果（ToTal F1-measure 73.1）都要优于ENet-Label-Torch的指标（Total F1-measure 72.0）。按照您给的信息，ENet-Label-Torch是ENet-SAD的方法，那么ERFNet-CULane-PyTorch对应的方法是否有加入SAD呢？如果没有加入SAD的ERFNet分割效果就优于ENet-SAD，那么在车道线检测任务上，一个合适的baseline网络如ERFNet，比增加一些耗时的操作如SCNN、或增加额外的工作如多任务网络、或使用一些训练技巧如您论文中提到的SAD，等等，都要简单且行之有效？

**Comment by cardwing** [comment]
目前的pytorch代码里并没有加入SAD过程。在车道线检测任务上，一个合理的基础网络确实很重要。你可以尝试用NAS来搜索一个网络，说不定效果会更好。不过，我之前也有用lua实现了基于ERFNet的车道线检测模型，但是效果比当前的pytorch版本差很多（F1-measure大概只有68左右）。所以，不同框架下的代码效果也有差异（可能一些函数功能不完全一样）。

**Comment by Baiwenjing** [comment]
谢谢作者分享，我也在车道线检测任务上摸索了一段时间，车道线有断续，因此它要求网络对上下文推理的要求更高一些。                         
另外，对于您的ERFNet-CULane-Pytorch中的args.no_partialbn参数的作用有些疑惑：
1. 大概猜测作者想通过它来控制是否想用“跨GPU的BN”，不知理解是否正确，具体是怎样通过配置实现呢？
2.    if args.no_partialbn:
        model.module.partialBN(False)   
        sync_bn.convert_bn(model)
    else:
        model.module.partialBN(True)   
请问这里的分支具体是什么含义？

3. 在class ERFNet中，有以下
if partial_bn:
       self.partialBN(True)
   …
   def train(self, mode=True):
        """
        Override the default train() to freeze the BN parameters
        :return:
        """
        super(ERFNet, self).train(mode)
        if self._enable_pbn:
            print("Freezing BatchNorm2D.")
            for m in self.modules():
                if isinstance(m, nn.BatchNorm2d):
                    m.eval()
                    # shutdown update in frozen mode
                    m.weight.requires_grad = False
                    m.bias.requires_grad = False

   def partialBN(self, enable):
        self._enable_pbn = enable

   这部分的代码实现的是什么含义及目的呢？

**Comment by cardwing** [comment]
args.no_partialbn参数确实是用来控制是否使用sync bn。代码的具体功能你自行跑一下代码就清楚了。

**Comment by Baiwenjing** [comment]
您好，还是想问一下，以下代码中，self._enable_pbn，意思是否是可以使用单GPU的BN更新，那为什么要freeze batchNorm2D呢？

    def train(self, mode=True):
        """
        Override the default train() to freeze the BN parameters
        :return:
        """
        super(ERFNet, self).train(mode)
        if self._enable_pbn:
            print("Freezing BatchNorm2D.")
            for m in self.modules():
                if isinstance(m, nn.BatchNorm2d):
                    m.eval()
                    # shutdown update in frozen mode
                    m.weight.requires_grad = False
                    m.bias.requires_grad = False

**Comment by Baiwenjing** [comment]
作者您好，还是有以下疑问：
1） 对于arg.no_partialbn = True，代表使用sync bn，这时 self._enable_pbn = False，
      那么，当arg.no_partialbn = False时，代表使用partial bn（单个GPU的BN）对吗？，这时   
      self._enable_pbn = True，那为什么要进行 freezing batchNorm2D的操作呢？这样在训练过程中，
      BN参数不就不能更新了啊，就不符合训练逻辑了吧？

**Comment by cardwing** [comment]
如果arg.no_partialbn = False，那么BN参数就固定住不更新，所有GPU共享一个BN参数。这就是不使用Sync BN的处理方式，不然每块GPU上的BN参数都不一样，最后训练完的网络取哪块GPU上的BN参数呢？并且，你可以试一下不使用Sync BN同时不固定BN参数。这样做很难训练网络的。

**Comment by Baiwenjing** [comment]
作者您好，感谢您的回复，按照您的解释：
============================================
“  如果arg.no_partialbn = False，那么BN参数就固定住不更新，所有GPU共享一个BN参数 ” 。
============================================
是否可以理解为所有的BN都不更新了，也就是说，所有GPU的BN参数值都是初始化给定的同一个初始值（可能是随机值），而在训练过程中都不参与训练、也不会更新。这样理解对吗？如果这样理解正确，那就是所有GPU都按照一个随机初始值做BN了，这样能保证效果吗？

**Comment by Baiwenjing** [comment]
作者您好，感谢您的耐心回复，按照您的解释：

“ 如果arg.no_partialbn = False，那么BN参数就固定住不更新，所有GPU共享一个BN参数 ” 。

是否可以理解为所有的BN都不更新了，也就是说，所有GPU的BN参数值都是初始化给定的同一个初始值（可能是随机值），而在训练过程中都不参与训练、也不会更新。这样理解对吗？如果这样理解正确，那就是所有GPU都按照一个随机初始值做BN了，这样能保证效果吗？

**Comment by cardwing** [comment]
是的。不过这里的BN参数是ImageNet上得到的，所以不更新不会有太大影响。但相比Sync BN，效果肯定会差一些。
